### PR TITLE
Add MODIS L2 products produced by IMAPP

### PR DESCRIPTION
**Pull request by djhoese**
As part of my work towards Polar2Grid 3.0, I needed to have other MODIS L2 products available in the `modis_l2` reader. We specifically support the files from IMAPP. There are some differences between these files and the NASA MODXX files so I've had to make some small adjustments to the python code. Otherwise this is mostly YAML additions.

 - [ ] Tests added

### CODE REVIEW

**Review by mraspaud — APPROVED** [review]
LGTM!

**Comment by djhoese** [comment]
@mraspaud @simonrp84 So here's the status of this PR:

1. Complete rewrite of all of the MODIS L2 tests and L1b tests. This is done by using pytest fixtures to create the files once per pytest session (they aren't recreated for every test and are assumed to only be read-only).
2. The pytest fixtures are added in a `satpy.tests.reader_tests._modis_fixtures` module which is imported in `satpy/tests/reader_tests/conftest.py`. They need to be used in `conftest.py` so pytest can make them available in multiple test modules.
3. Added tests for mod35 and mod06 files.
4. Added YAML configuration for IMAPP output that we use in Polar2Grid (except for a few, see TODO below)
5. Added YAML configuration so that L2 files that don't have SensorZenith data can still interpolate geolocation from 1km to 500m or 250m using the new version of python-geotiepoints.

TODO:

- [ ] Add `cloud_mask`, `land_sea_mask`, and `snow_ice_mask` from the IMAPP `mask_byte1` file
- [ ] Add test cases for the other L2 file types

If you have some free time then reviewing what's here already would be appreciated.

**Comment by djhoese** [comment]
@mraspaud This PR is ready for review and merge. Some of the handling of _FillValue and masks and preserving integer data types is not as clean as I would have hoped, but this is mostly due to the MODIS files themselves. Some files I've encountered have a scale_factor of 1.0 and add_offset of 0.0, but are actually not category products and are meant to be treated as floats. Otherwise, before this PR the readers didn't preserve integer types for category types at all.

If you'd like to schedule a video call to "pair program"/review this then I can do that. Let me know.

File: satpy/readers/modis_l2.py [hunk 1]
```diff
@@ -104,112 +109,116 @@ def read_geo_resolution(metadata):
...
+                              for dataset_dim, quality_assurance_dim
+                              in zip(dataset.shape, quality_assurance.shape)]
+        quality_assurance = np.tile(quality_assurance, duplication_factor)
+        # Replace unassured data by NaN value
+        dataset[np.where(quality_assurance == 0)] = dataset.attrs["_FillValue"]
+        return dataset
+
 
-def bits_strip(bit_start, bit_count, value):
+def _extract_byte_mask(dataset, byte_information, bit_start, bit_count):
```
Comment (mraspaud) [inline satpy/readers/modis_l2.py:196]: There is also a bit mask extraction class in the olci reader, maybe we could factorize these two in the future?

**Review by mraspaud — APPROVED** [review]
LGTM! thanks for adding all the tests!

File: satpy/readers/modis_l2.py (hunk 1 above)
Comment (djhoese) [reply satpy/readers/modis_l2.py:196]: Yes we could probably. I didn't actually implement any of this I just extracted it from the method in the file handler. I'm sure there are some small differences, but I have no feelings about factorizing in the future.

**Comment by djhoese** [comment]
I know @simonrp84 wanted to do some testing. I'm going to do some of the basic things that I think he's been doing like loading a `true_color` and seeing what happens.

**Comment by djhoese** [comment]
Ok so I was able to take some NASA L1b files (all 3 data and 1 geo) and do `scn.load(['true_color'])` plus resampling to get the result and I was able to do `scn.load(['true_color'], resolution=1000)` without resampling and that worked. @simonrp84 any other ideas for tests?

**Comment by djhoese** [comment]
I think I'm going to merge this due to the urgency of #1816 and #1823 needing some of the fixes in this PR. I will make PRs for any future fixes, but most things seem to work and my tests are fairly decent if I do say so myself. :wink:

**Comment by simonrp84** [comment]
Just gave this a quick test with the L1 data I have to hand and I didn't encounter any problems. Far from a robust test, but LGTM!

**Pull request merged**